### DISCUSSION TITLE
Data types

### DISCUSSION THREAD

**Discussion by deltamarnix** [pyphoenix/pyphoenix-project#54] [Ideas]
The type hierarchy for data types in flopy3 is quite complicated. It would be better if we could use builtin or 3rd party data types and have their auxiliary functions situated somewhere else.

![image](https://github.com/user-attachments/assets/69758ba8-26be-439f-bf78-1bf5b591ec9c)

This ties in with #49 for grid data types.

**Comment by wpbonelli** [comment]
Maybe this can be concluded since it is covered by #49 and #67.

Tentatively, we've discussed

- component variables as `attrs.field`s type-hinted as builtin primitives/collections, numpy arrays, or composites of such
- every component having a `data` attribute serving as its data store, which is either an `xarray.DataSet` or `DataTree`, depending whether the component has children &mdash; the `attrs` attributes proxy the xarray store rather than storing values themselves

So users can interact with component variables directly like `dis.delr` as well as xarray-style `dis.data["delr"]`, and changes in either will be reflected in the other.

And we get a lot for free from xarray, hopefully making all those util modules and data structures you show unnecessary.

Regarding transient data, isn't it just an array with an extra time dimension, indexed by kper/kstp?

**Reply by wpbonelli** [reply]
if we end up adopting (some subset of) CF conventions https://github.com/xarray-contrib/cf-xarray

**Comment by wpbonelli** [comment]
Regarding keyword options (i.e. booleans): https://github.com/modflowpy/flopy/discussions/2454

Assuming all defaults are defined in DFNs, flopy could simply never use `Optional[bool]`, just propagate the default and write all keyword options out in the interest of transparency/explicitness (or this could be toggled with a setting)? Assuming MF6 never distinguishes the absence of a keyword variable from an explicit false value. I don't think it does?
